python/paddle/distributed/auto_parallel/static/tuner/pir_rule_based_tuner.py

```diff
@@ -0,0 +1,1487 @@
+                sin = sin[:, : q.shape[1], :, :]  # [bs, seq_len, 1, dim]
+            else:
+                cos = cos.squeeze(axis=[0, 2])  # [seq_len, dim]
+                sin = sin.squeeze(axis=[0, 2])  # [seq_len, dim]
+                cos = cos[position_ids].unsqueeze(2)  # [bs, seq_len, 1, dim]
+                sin = sin[position_ids].unsqueeze(2)  # [bs, seq_len, 1, dim]
+            q_embed = (q * cos) + (self.rotate_half(q) * sin)
+            k_embed = (k * cos) + (self.rotate_half(k) * sin)
+
+            # scale_dot_product
+            tmp = self.scale_dot_product_attention(
+                q_embed, k_embed, v, attention_mask
+            )
+
+            # out_linear
+            output = paddle.matmul(tmp, out_weight)
+
+        self.pir_program = main_program
+        # print(f"in AttentionPattern, program is {self.pir_program}")
+        paddle.disable_static()
+
+        # # todo: how to design an efficient distributed infos for each pattern
+        qkv_linear_dist_infos = MpDistInfos("column")
+        out_linear_dist_infos = MpDistInfos("row")
+
+        # # # build ops dist infos # # #
+        ops_dist_infos = {
+            (9,): qkv_linear_dist_infos,
+            (10,): qkv_linear_dist_infos,
+            (11,): qkv_linear_dist_infos,
+            (77,): out_linear_dist_infos,
+        }
+        self.ops_dist_infos = ops_dist_infos
+
+    @staticmethod
+    def rotate_half(x):
+        x1 = x[..., : x.shape[-1] // 2]
+        x2 = x[..., x.shape[-1] // 2 :]
+        return paddle.concat([-x2, x1], axis=-1)  # shape is the same as x
+
+    @staticmethod
+    def scale_dot_product_attention(
+        query_states,
+        key_states,
+        value_states,
+        attention_mask,
+    ):
+
+        bsz, q_len, num_heads, head_dim = query_states.shape
+        _, kv_seq_len, _, _ = value_states.shape
+
+        #  [ bz, seqlen, nhead, head_dim] -> [bs, nhead, seq_len, head_dim]
+        query_states = paddle.transpose(query_states, [0, 2, 1, 3])
+        # merge with the next tranpose
+        key_states = paddle.transpose(key_states, [0, 2, 1, 3])
+        value_states = paddle.transpose(value_states, [0, 2, 1, 3])
+
+        # matmul and devide by sqrt(head_dim)
+        attn_weights = paddle.matmul(
+            query_states / math.sqrt(head_dim),
+            key_states.transpose([0, 1, 3, 2]),
+        )
+
+        attention_mask = attention_mask.reshape([bsz, 1, q_len, kv_seq_len])
+
+        attn_weights = attn_weights + attention_mask
+
+        attn_weights = F.softmax(attn_weights, axis=-1, dtype="float32").astype(
+            query_states.dtype
+        )
+
+        attn_output = paddle.matmul(attn_weights, value_states)
+        attn_output = attn_output.transpose([0, 2, 1, 3])
+
+        attn_output = attn_output.reshape([bsz, q_len, head_dim * num_heads])
+
+        return attn_output
+
+
+# Llama
+@register_pir_pattern
+class PIRMLPPattern(PIRBasePattern):
```

Comment:
There are many variations of `MLP`, such as having 2 or 3 linear layers, with bias or no bias, and the activation function may also be different, so the name of `MLPPattern` needs to be carefully considered.

Reply:
also, is it necessary to use the prefix "PIR"?

Reply:
given that pir as default feature for paddlepaddle, remove all prefix "PIR" here